sys-cluster/hpx/hpx-1.5.1.ebuild

```diff
@@ -0,0 +1,119 @@
+# Copyright 1999-2020 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+PYTHON_COMPAT=( python3_{6..9} )
+
+if [[ ${PV} == 9999 ]] ; then
+	inherit git-r3
+	EGIT_REPO_URI="https://github.com/STEllAR-GROUP/hpx.git"
+else
+	SRC_URI="https://github.com/STEllAR-GROUP/${PN}/archive/${PV}.tar.gz -> ${P}.tar.gz"
+	KEYWORDS="~amd64 ~x86 ~amd64-linux ~x86-linux"
+fi
+inherit check-reqs cmake multiprocessing python-single-r1
+
+DESCRIPTION="C++ runtime system for parallel and distributed applications"
+HOMEPAGE="https://stellar.cct.lsu.edu/tag/hpx/"
+
+SLOT="0"
+LICENSE="Boost-1.0"
+IUSE="doc examples jemalloc mpi papi +perftools tbb test"
+RESTRICT="!test? ( test )"
+
+REQUIRED_USE="
+	${PYTHON_REQUIRED_USE}
+	?? ( jemalloc perftools tbb )
+"
+
+BDEPEND="
+	virtual/pkgconfig
+	doc? (
+		${PYTHON_DEPS}
+		app-doc/doxygen
+		$(python_gen_cond_dep '
+			dev-python/sphinx[${PYTHON_MULTI_USEDEP}]
+			dev-python/sphinx_rtd_theme[${PYTHON_MULTI_USEDEP}]
+			>=dev-python/breathe-4.22[${PYTHON_MULTI_USEDEP}]
+		')
+	)
+	test? ( ${PYTHON_DEPS} )
+"
+RDEPEND="
+	${PYTHON_DEPS}
+	dev-libs/boost:=
+	sys-apps/hwloc
+	sys-libs/zlib
+	mpi? ( virtual/mpi )
+	papi? ( dev-libs/papi )
+	perftools? ( dev-util/google-perftools )
+	tbb? ( dev-cpp/tbb )
+"
+DEPEND="${RDEPEND}"
+
+PATCHES=(
+	"${FILESDIR}/${P}-cmake.patch"
+	"${FILESDIR}/${P}-python.patch"
+	"${FILESDIR}/${P}-tests.patch"
+)
+
+hpx_memory_requirement() {
+	# HPX needs enough main memory for compiling
+	# rule of thumb: 1G per job
+	if [[ -z ${MAKEOPTS} ]] ; then
+		echo "2G"
+	else
+		local jobs=$(makeopts_jobs "${MAKEOPTS}" "$(get_nproc)")
+		echo "${jobs}G"
+	fi
+}
+
+pkg_pretend() {
+	local CHECKREQS_MEMORY=$(hpx_memory_requirement)
+	check-reqs_pkg_setup
+}
+
+pkg_setup() {
+	local CHECKREQS_MEMORY=$(hpx_memory_requirement)
+	check-reqs_pkg_setup
+	python-single-r1_pkg_setup
+}
+
+src_configure() {
+	local mycmakeargs=(
+		-DHPX_WITH_EXAMPLES=OFF
+		-DHPX_WITH_DOCUMENTATION=$(usex doc)
```

Comment:
```
/var/tmp/portage/gpm/portage/sys-cluster/hpx-1.5.1/work/hpx-1.5.1_build/docs/sphinx/libs/futures/docs/index.rst:14: WARNING: Unparseable C++ cross-reference: 'hpx:lcos::shared_future'
Error in cross-reference.
If shorthand ref:
  Invalid C++ declaration: Expected end of definition. [error at 3]
    hpx:lcos::shared_future
    ---^
If full function ref:
  Error when parsing function declaration.
  If the function has no return type:
    Error in declarator or parameters-and-qualifiers
    Invalid C++ declaration: Expecting "(" in parameters-and-qualifiers. [error at 3]
      hpx:lcos::shared_future
      ---^
  If the function has a return type:
    Error in declarator or parameters-and-qualifiers
    If pointer to member declarator:
      Invalid C++ declaration: Expected identifier in nested name. [error at 3]
        hpx:lcos::shared_future
        ---^
    If declarator-id:
      Invalid C++ declaration: Expected identifier in nested name. [error at 3]
        hpx:lcos::shared_future
        ---^

Exception occurred:
  File "/usr/lib/python3.8/site-packages/sphinx/domains/cpp.py", line 7375, in _resolve_xref_inner
    assert parentSymbol  # should be there
AssertionError
The full traceback has been saved in /var/tmp/portage/gpm/portage/sys-cluster/hpx-1.5.1/temp/sphinx-err-sqyorrt_.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!

ninja: build stopped: subcommand failed.
 * ERROR: sys-cluster/hpx-1.5.1::testworld failed (compile phase):
```
```
 >: cat /var/tmp/portage/gpm/portage/sys-cluster/hpx-1.5.1/temp/sphinx-err-sqyorrt_.log
# Sphinx version: 3.2.1
# Python version: 3.8.6 (CPython)
# Docutils version: 0.16 release
# Jinja2 version: 2.11.2
# Last messages:
#   writing output... [ 64%] libs/runtime_configuration/docs/index
#   writing output... [ 64%] libs/runtime_local/api
#   writing output... [ 65%] libs/runtime_local/docs/index
#   writing output... [ 65%] libs/schedulers/api
#   writing output... [ 66%] libs/schedulers/docs/index
#   writing output... [ 66%] libs/segmented_algorithms/api
#   writing output... [ 67%] libs/segmented_algorithms/docs/index
#   writing output... [ 67%] libs/serialization/api
#   writing output... [ 67%] libs/serialization/docs/index
#   writing output... [ 68%] libs/static_reinit/api
# Loaded extensions:
#   sphinx.ext.mathjax (3.2.1) from /usr/lib/python3.8/site-packages/sphinx/ext/mathjax.py
#   sphinxcontrib.applehelp (1.0.2) from /usr/lib/python3.8/site-packages/sphinxcontrib/applehelp/__init__.py
#   sphinxcontrib.devhelp (1.0.2) from /usr/lib/python3.8/site-packages/sphinxcontrib/devhelp/__init__.py
#   sphinxcontrib.htmlhelp (1.0.3) from /usr/lib/python3.8/site-packages/sphinxcontrib/htmlhelp/__init__.py
#   sphinxcontrib.serializinghtml (1.1.4) from /usr/lib/python3.8/site-packages/sphinxcontrib/serializinghtml/__init__.py
#   sphinxcontrib.qthelp (1.0.3) from /usr/lib/python3.8/site-packages/sphinxcontrib/qthelp/__init__.py
#   alabaster (0.7.12) from /usr/lib/python3.8/site-packages/alabaster/__init__.py
#   sphinx.ext.autodoc.type_comment (3.2.1) from /usr/lib/python3.8/site-packages/sphinx/ext/autodoc/type_comment.py
#   sphinx.ext.autodoc (3.2.1) from /usr/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py
#   sphinx.ext.doctest (3.2.1) from /usr/lib/python3.8/site-packages/sphinx/ext/doctest.py
#   sphinx.ext.intersphinx (3.2.1) from /usr/lib/python3.8/site-packages/sphinx/ext/intersphinx.py
#   sphinx.ext.todo (3.2.1) from /usr/lib/python3.8/site-packages/sphinx/ext/todo.py
#   sphinx.ext.coverage (3.2.1) from /usr/lib/python3.8/site-packages/sphinx/ext/coverage.py
#   sphinx.ext.ifconfig (3.2.1) from /usr/lib/python3.8/site-packages/sphinx/ext/ifconfig.py
#   sphinx.ext.viewcode (3.2.1) from /usr/lib/python3.8/site-packages/sphinx/ext/viewcode.py
#   breathe (4.22.1) from /usr/lib/python3.8/site-packages/breathe/__init__.py
#   sphinx-hpx (unknown version) from /var/tmp/portage/gpm/portage/sys-cluster/hpx-1.5.1/work/hpx-1.5.1_build/docs/sphinx/extensions/sphinx-hpx.py
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/sphinx/cmd/build.py", line 280, in build_main
    app.build(args.force_all, filenames)
  File "/usr/lib/python3.8/site-packages/sphinx/application.py", line 348, in build
    self.builder.build_update()
  File "/usr/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 297, in build_update
    self.build(to_build,
  File "/usr/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 361, in build
    self.write(docnames, list(updated_docnames), method)
  File "/usr/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 535, in write
    self._write_serial(sorted(docnames))
  File "/usr/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 542, in _write_serial
    doctree = self.env.get_and_resolve_doctree(docname, self)
  File "/usr/lib/python3.8/site-packages/sphinx/environment/__init__.py", line 539, in get_and_resolve_doctree
    self.apply_post_transforms(doctree, docname)
  File "/usr/lib/python3.8/site-packages/sphinx/environment/__init__.py", line 585, in apply_post_transforms
    transformer.apply_transforms()
  File "/usr/lib/python3.8/site-packages/sphinx/transforms/__init__.py", line 87, in apply_transforms
    super().apply_transforms()
  File "/usr/lib/python3.8/site-packages/docutils/transforms/__init__.py", line 171, in apply_transforms
    transform.apply(**kwargs)
  File "/usr/lib/python3.8/site-packages/sphinx/transforms/post_transforms/__init__.py", line 44, in apply
    self.run(**kwargs)
  File "/usr/lib/python3.8/site-packages/sphinx/transforms/post_transforms/__init__.py", line 87, in run
    newnode = domain.resolve_xref(self.env, refdoc, self.app.builder,
  File "/usr/lib/python3.8/site-packages/sphinx/domains/cpp.py", line 7501, in resolve_xref
    return self._resolve_xref_inner(env, fromdocname, builder, typ,
  File "/usr/lib/python3.8/site-packages/sphinx/domains/cpp.py", line 7375, in _resolve_xref_inner
    assert parentSymbol  # should be there
AssertionError
```
Seems to be doc-related. Rest looks good.

Reply:
Thanks for the review. Yes, this is doc related. I'll debug it.

Reply:
Pushed a fix for it. Thanks.